File: LeetCode/Easy/1290-Convert_Binary_Number_In_A_Linked_List_To_Integer/1290-ConvertBinaryNumberInALinkedListToInteger.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end

defmodule Solution do
  @spec get_decimal_value(head :: ListNode.t | nil) :: integer
  def get_decimal_value(head) do
    solve(head, 0)
  end

  defp solve(nil, num), do: num

  defp solve(curr, num) do
    solve(curr.next, (num * 2) + curr.val)
  end
end

File: LeetCode/Easy/1290-Convert_Binary_Number_In_A_Linked_List_To_Integer/1290-ConvertBinaryNumberInALinkedListToInteger.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    int getDecimalValue(ListNode *head)
    {
        int num = 0;
        ListNode *curr = head;

        while (curr != nullptr)
        {
            num = (num << 1) + curr->val;
            curr = curr->next;
        }

        return num;
    }
};